examples/mnist.py

```diff
@@ -33,12 +33,28 @@ def load_dataset():
     # We first define a download function, supporting both Python 2 and 3.
     if sys.version_info[0] == 2:
         from urllib import urlretrieve
+        URLError = IOError
     else:
-        from urllib.request import urlretrieve
-
-    def download(filename, source='http://yann.lecun.com/exdb/mnist/'):
+        from urllib.request import urlretrieve, URLError
+
+    def download(filename,
+                 sources=['http://yann.lecun.com/exdb/mnist/',
+                          'https://web.archive.org/web/20150906081542/'
+                          'http://yann.lecun.com/exdb/mnist/',
+                          'https://s3.amazonaws.com/lasagne/recipes/'
+                          'datasets/mnist/']):
         print("Downloading %s" % filename)
-        urlretrieve(source + filename, filename)
+        for source in sources:
+            try:
+                urlretrieve(source + filename, filename)
+            except URLError as exc:
+                if exc.args[-1].args == (-2, 'Name or service not known'):
+                    sources.remove(source)  # do not try this source again
+            else:
+                break  # we were successful!
+        else:
+            print("Could not download from any location. Are you offline?")
```

Comment:
I figured it would be easier to spot if it comes right after the "Downloading %s" line, instead of being mixed in somewhere with the exception backtrace.